File: lib/presentation/views/circles/gratitude_wall_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../../domain/entities/circle.dart';
import '../../../domain/repositories/circle_repository.dart';
import '../../theme/app_theme.dart';

/// Embeddable gratitude wall widget (used inside CircleDetailView).
class GratitudeWallWidget extends StatefulWidget {
  final String circleId;
  const GratitudeWallWidget({super.key, required this.circleId});

  @override
  State<GratitudeWallWidget> createState() => _GratitudeWallWidgetState();
}

class _GratitudeWallWidgetState extends State<GratitudeWallWidget> {
  List<GratitudePost> _gratitudes = [];
  bool _isLoading = true;
  bool _isLoadingMore = false;
  bool _hasOlderWeeks = true;
  int _weeksBack = 0;
  int _newCount = 0;
  bool _showNewBadge = false;
  bool _loadError = false;
  GratitudePost? _deleteTarget;

  @override
  void initState() {
    super.initState();
    _loadWall();
    _loadNewCount();
    _markSeen();
  }

  Future<void> _loadWall() async {
    setState(() { _isLoading = true; _loadError = false; });
    try {
      final r = await context.read<CircleRepository>().getGratitudeWall(widget.circleId, weeksBack: _weeksBack);
      if (mounted) setState(() { _gratitudes = r.gratitudes; _isLoading = false; });
    } catch (_) {
      if (mounted) setState(() { _isLoading = false; _loadError = true; });
    }
  }

  Future<void> _loadNewCount() async {
    try {
      final count = await context.read<CircleRepository>().getGratitudeNewCount(widget.circleId);
      if (mounted && count > 0) {
        setState(() { _newCount = count; _showNewBadge = true; });
        Future.delayed(const Duration(seconds: 5), () {
          if (mounted) setState(() => _showNewBadge = false);
        });
      }
    } catch (_) {}
  }

  Future<void> _markSeen() async {
    try { await context.read<CircleRepository>().markGratitudesSeen(widget.circleId); } catch (_) {}
  }

  void _loadPreviousWeek() {
    setState(() => _isLoadingMore = true);
    final nextWeek = _weeksBack + 1;
    context.read<CircleRepository>().getGratitudeWall(widget.circleId, weeksBack: nextWeek).then((r) {
      if (!mounted) return;
      if (r.gratitudes.isEmpty) {
        setState(() { _hasOlderWeeks = false; _isLoadingMore = false; });
      } else {
        setState(() {
          _gratitudes.addAll(r.gratitudes);
          _weeksBack = nextWeek;
          _isLoadingMore = false;
        });
      }
    }).catchError((_) {
      if (mounted) setState(() => _isLoadingMore = false);
    });
  }

  Future<void> _deleteGratitude(GratitudePost item) async {
    try {
      await context.read<CircleRepository>().deleteGratitude(widget.circleId, item.id);
      if (mounted) setState(() => _gratitudes.removeWhere((g) => g.id == item.id));
    } catch (_) {}
  }

  String _relativeTime(String dateString) {
    final date = DateTime.tryParse(dateString);
    if (date == null) return '';
    final diff = DateTime.now().difference(date);
    if (diff.inMinutes < 1) return 'Just now';
    if (diff.inHours < 1) return '${diff.inMinutes}m ago';
    if (diff.inDays < 1) return '${diff.inHours}h ago';
    if (diff.inDays == 1) return 'Yesterday';
    final days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    return days[(date.weekday - 1) % 7];
  }

  @override
  Widget build(BuildContext context) {
    return Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
      Text('GRATITUDE WALL',
          style: TextStyle(fontSize: 11, fontWeight: FontWeight.w600,
              color: Colors.white.withValues(alpha: 0.4), letterSpacing: 1.2)),
      const SizedBox(height: 10),
      if (_isLoading)
        const Center(child: Padding(
          padding: EdgeInsets.all(16),
          child: CircularProgressIndicator(color: MyWalkColor.golden, strokeWidth: 2),
        ))
      else if (_loadError && _gratitudes.isEmpty)
        Padding(
          padding: const EdgeInsets.symmetric(vertical: 8),
          child: Center(
            child: Text(
              "Couldn't connect. Check your connection.",
              style: TextStyle(fontSize: 13, color: MyWalkColor.warmCoral.withValues(alpha: 0.8)),
            ),
          ),
        )
      else if (_gratitudes.isEmpty)
        Padding(
          padding: const EdgeInsets.symmetric(vertical: 8),
          child: Center(
            child: Text(
              _weeksBack == 0 ? 'No gratitudes shared this week yet' : 'No gratitudes this week',
              style: TextStyle(fontSize: 13, color: Colors.white.withValues(alpha: 0.4)),
            ),
          ),
        )
      else ...[
        if (_showNewBadge && _newCount > 0)
          Center(
            child: Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: Text('$_newCount new gratitude${_newCount == 1 ? '' : 's'}',
                  style: const TextStyle(fontSize: 12, fontWeight: FontWeight.w500, color: MyWalkColor.golden)),
            ),
          ),
        ..._gratitudes.map((item) => Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: _gratitudeCard(item),
            )),
        if (_hasOlderWeeks)
          Center(
            child: TextButton(
              onPressed: _isLoadingMore ? null : _loadPreviousWeek,
              child: _isLoadingMore
                  ? const SizedBox(width: 16, height: 16,
                      child: CircularProgressIndicator(strokeWidth: 2, color: MyWalkColor.golden))
                  : Text('Previous weeks',
                      style: TextStyle(fontSize: 12, color: MyWalkColor.golden.withValues(alpha: 0.7))),
            ),
          ),
      ],
      if (_deleteTarget != null)
        _confirmDeleteDialog(),
    ]);
  }

  Widget _gratitudeCard(GratitudePost item) {
    return GestureDetector(
      onLongPress: item.isMine
          ? () => setState(() => _deleteTarget = item)
          : null,
      child: Container(
        padding: const EdgeInsets.fromLTRB(16, 14, 16, 14),
        decoration: BoxDecoration(
          color: const Color(0xFF2A2A3C),
          borderRadius: BorderRadius.circular(12),
          border: Border.all(color: const Color(0xFF353548), width: 1),
        ),
        child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
          Row(children: [
            Text(
              item.isAnonymous
                  ? 'Someone in your circle'
                  : (item.displayName?.split(' ').first ?? 'Member'),
              style: TextStyle(
                  fontSize: 13, fontWeight: FontWeight.w600,
                  color: item.isAnonymous
                      ? const Color(0xFF9A98A0)
                      : MyWalkColor.warmWhite),
            ),
            const Spacer(),
            Text(_relativeTime(item.sharedAt),
                style: const TextStyle(fontSize: 11, color: Color(0xFF6B6B7B))),
          ]),
          const SizedBox(height: 6),
          Text(item.gratitudeText,
              style: const TextStyle(fontSize: 14, color: MyWalkColor.warmWhite, height: 1.5)),
        ]),
      ),
    );
  }

  Widget _confirmDeleteDialog() {
    return AlertDialog(
      backgroundColor: MyWalkColor.cardBackground,
      title: const Text('Delete Gratitude', style: TextStyle(color: MyWalkColor.warmWhite)),
      content: Text(
        'This gratitude will be removed from the wall for all circle members.',
        style: TextStyle(color: Colors.white.withValues(alpha: 0.6)),
      ),
      actions: [
        TextButton(
          onPressed: () => setState(() => _deleteTarget = null),
          child: Text('Cancel', style: TextStyle(color: Colors.white.withValues(alpha: 0.5))),
        ),
        TextButton(
          onPressed: () {
            final t = _deleteTarget;
            setState(() => _deleteTarget = null);
            if (t != null) _deleteGratitude(t);
          },
          child: const Text('Delete', style: TextStyle(color: MyWalkColor.warmCoral)),
        ),
      ],
    );
  }
}

/// Full-screen gratitude wall view (standalone navigation target).
class GratitudeWallView extends StatelessWidget {
  final String circleId;
  const GratitudeWallView({super.key, required this.circleId});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: MyWalkColor.charcoal,
      appBar: AppBar(
        backgroundColor: MyWalkColor.charcoal,
        title: const Text('Gratitude Wall',
            style: TextStyle(color: MyWalkColor.warmWhite, fontSize: 20, fontWeight: FontWeight.w700)),
      ),
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(16),
          child: GratitudeWallWidget(circleId: circleId),
        ),
      ),
    );
  }
}

File: swift_reference/Views/GratitudeWallView.swift
import SwiftUI

struct GratitudeWallView: View {
    let circleId: String

    @State private var gratitudes: [SharedGratitudeItem] = []
    @State private var isLoading: Bool = true
    @State private var weeksBack: Int = 0
    @State private var isLoadingMore: Bool = false
    @State private var hasOlderWeeks: Bool = true
    @State private var newCount: Int = 0
    @State private var showNewBadge: Bool = false
    @State private var deleteTarget: SharedGratitudeItem?

    var body: some View {
        Section {
            if isLoading {
                HStack {
                    Spacer()
                    ProgressView()
                        .controlSize(.small)
                    Spacer()
                }
                .listRowBackground(Color.clear)
            } else if gratitudes.isEmpty {
                HStack {
                    Spacer()
                    Text(weeksBack == 0 ? "No gratitudes shared this week yet" : "No gratitudes this week")
                        .font(.system(.caption, design: .serif))
                        .foregroundStyle(Color(hex: "9A98A0"))
                    Spacer()
                }
                .padding(.vertical, 8)
                .listRowBackground(Color.clear)
            } else {
                if showNewBadge && newCount > 0 {
                    HStack {
                        Spacer()
                        Text("\(newCount) new gratitude\(newCount == 1 ? "" : "s")")
                            .font(.system(size: 12, weight: .medium))
                            .foregroundStyle(TributeColor.golden)
                        Spacer()
                    }
                    .listRowBackground(Color.clear)
                    .listRowSeparator(.hidden)
                }

                ForEach(gratitudes) { item in
                    gratitudeCard(item)
                        .listRowBackground(Color.clear)
                        .listRowSeparator(.hidden)
                        .listRowInsets(EdgeInsets(top: 4, leading: 16, bottom: 4, trailing: 16))
                        .contextMenu {
                            if item.isMine {
                                Button(role: .destructive) {
                                    deleteTarget = item
                                } label: {
                                    Label("Delete", systemImage: "trash")
                                }
                            }
                        }
                }

                if hasOlderWeeks {
                    Button {
                        loadPreviousWeek()
                    } label: {
                        HStack {
                            Spacer()
                            if isLoadingMore {
                                ProgressView()
                                    .controlSize(.small)
                            } else {
                                Text("Previous weeks")
                                    .font(.caption)
                                    .foregroundStyle(TributeColor.golden.opacity(0.7))
                            }
                            Spacer()
                        }
                    }
                    .disabled(isLoadingMore)
                    .listRowBackground(Color.clear)
                    .listRowSeparator(.hidden)
                }
            }
        } header: {
            Text("GRATITUDE WALL")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(Color(hex: "6B6B7B"))
                .tracking(1.2)
        }
        .task {
            await loadWall()
            await loadNewCount()
            await markSeen()
        }
        .alert("Delete Gratitude", isPresented: Binding(
            get: { deleteTarget != nil },
            set: { if !$0 { deleteTarget = nil } }
        )) {
            Button("Delete", role: .destructive) {
                if let target = deleteTarget {
                    performDelete(target)
                }
            }
            Button("Cancel", role: .cancel) {
                deleteTarget = nil
            }
        } message: {
            Text("This gratitude will be removed from the wall for all circle members.")
        }
        .onAppear {
            startNewBadgeTimer()
        }
    }

    private func gratitudeCard(_ item: SharedGratitudeItem) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                if item.isAnonymous {
                    Text("Someone in your circle")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Color(hex: "9A98A0"))
                } else {
                    Text(item.displayName?.components(separatedBy: " ").first ?? "Member")
                        .font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(TributeColor.warmWhite)
                }

                Spacer()

                Text(relativeTime(item.sharedAt))
                    .font(.system(size: 11))
                    .foregroundStyle(Color(hex: "6B6B7B"))
            }

            Text(item.gratitudeText)
                .font(.system(size: 14))
                .foregroundStyle(TributeColor.warmWhite)
                .lineSpacing(4)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
        .background(Color(hex: "2A2A3C"))
        .clipShape(.rect(cornerRadius: 12))
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .strokeBorder(Color(hex: "353548"), lineWidth: 1)
        )
    }

    private func relativeTime(_ dateString: String) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        var date = formatter.date(from: dateString)
        if date == nil {
            let fallback = ISO8601DateFormatter()
            date = fallback.date(from: dateString)
        }
        guard let date else { return "" }

        let now = Date()
        let interval = now.timeIntervalSince(date)

        if interval < 60 { return "Just now" }
        if interval < 3600 { return "\(Int(interval / 60))m ago" }
        if interval < 86400 { return "\(Int(interval / 3600))h ago" }

        let calendar = Calendar.current
        if calendar.isDateInYesterday(date) { return "Yesterday" }

        let weekday = DateFormatter()
        weekday.dateFormat = "EEE"
        return weekday.string(from: date)
    }

    private func loadWall() async {
        isLoading = true
        do {
            let response = try await APIService.shared.getGratitudeWall(circleId: circleId, weeksBack: weeksBack)
            gratitudes = response.gratitudes
        } catch {}
        isLoading = false
    }

    private func loadNewCount() async {
        do {
            let response = try await APIService.shared.getGratitudeNewCount(circleId: circleId)
            newCount = response.newCount
        } catch {}
    }

    private func markSeen() async {
        do {
            try await APIService.shared.markGratitudesSeen(circleId: circleId)
        } catch {}
    }

    private func loadPreviousWeek() {
        isLoadingMore = true
        let nextWeek = weeksBack + 1
        Task {
            do {
                let response = try await APIService.shared.getGratitudeWall(circleId: circleId, weeksBack: nextWeek)
                if response.gratitudes.isEmpty {
                    hasOlderWeeks = false
                } else {
                    gratitudes.append(contentsOf: response.gratitudes)
                    weeksBack = nextWeek
                }
            } catch {}
            isLoadingMore = false
        }
    }

    private func performDelete(_ item: SharedGratitudeItem) {
        Task {
            do {
                try await APIService.shared.deleteGratitude(circleId: circleId, gratitudeId: item.id)
                withAnimation {
                    gratitudes.removeAll { $0.id == item.id }
                }
            } catch {}
        }
    }

    private func startNewBadgeTimer() {
        guard newCount > 0 else { return }
        showNewBadge = true
        Task {
            try? await Task.sleep(for: .seconds(5))
            withAnimation(.easeOut(duration: 0.3)) {
                showNewBadge = false
            }
        }
    }
}
